inc/media_offload.php

```diff
@@ -192,11 +218,83 @@ public static function instance() {
 				add_filter( 'media_row_actions', [self::$instance, 'add_inline_media_action'], 10, 2 );
 				add_filter( 'wp_calculate_image_srcset', [self::$instance, 'calculate_image_srcset'], 1, 5 );
 				add_action( 'post_updated', [self::$instance, 'update_offload_meta'], 10, 3 );
+				add_filter( 'wp_insert_attachment_data', [self::$instance, 'insert'], 10, 4 );
 			}
 		}
 		return self::$instance;
 	}
 
+
+	/**
+	 * Function for `update_attached_file` filter-hook.
+	 *
+	 * @param string $file          Path to the attached file to update.
+	 * @param int    $attachment_id Attachment ID.
+	 *
+	 * @return string
+	 */
+	function wp_update_attached_file_filter( $file, $attachment_id ) {
+
+		if ( OPTML_DEBUG_MEDIA ) {
+			do_action( 'optml_log', 'called updated attached' );
+		}
+		$info = pathinfo( $file );
+		$file_name = basename( $file );
+		$no_ext_file_name = basename( $file, '.' . $info['extension'] );
+
+		if ( ! empty( self::$current_file_deduplication ) && stripos( self::$current_file_deduplication, $no_ext_file_name ) !== false ) {
+			$file = str_replace( $file_name, self::$current_file_deduplication, $file );
+			self::$last_deduplicated = strtolower( $file_name );
+			self::$last_deduplicated_original = $file_name;
+			self::$current_file_deduplication = false;
+		}
+		if ( OPTML_DEBUG_MEDIA ) {
+			do_action( 'optml_log', self::$last_deduplicated );
+		}
+		return $file;
+	}
+
+	/**
+	 * Function for `wp_insert_attachment_data` filter-hook.
+	 *
+	 * @param array $data                An array of slashed, sanitized, and processed attachment post data.
+	 * @param array $postarr             An array of slashed and sanitized attachment post data, but not processed.
+	 * @param array $unsanitized_postarr An array of slashed yet *unsanitized* and unprocessed attachment post data as originally passed to wp_insert_post().
+	 * @param bool  $update              Whether this is an existing attachment post being updated.
+	 *
+	 * @return array
+	 */
+	function insert( $data, $postarr, $unsanitized_postarr, $update ) {
+
+		// the post name is unique against the database so not affected by removing the files
+		// https://developer.wordpress.org/reference/functions/wp_unique_post_slug/
+		if ( OPTML_DEBUG_MEDIA ) {
+			do_action( 'optml_log', 'data before' );
+			do_action( 'optml_log', $data );
+		}
+		if ( ! empty( $data['guid'] ) ) {
+			$filename = wp_basename( $data['guid'] );
+			$ext = $this->get_ext( $filename );
+			$sanitized_post_name = str_replace( '-' . $ext, '', $data['post_name'] );
+		}
+		if ( ! empty( $data['post_name'] ) && $data['post_title'] !== $sanitized_post_name ) {
+			$to_replace_with = $sanitized_post_name . '.' . $ext;
+			$data['guid'] = str_replace( $filename, $to_replace_with, $data['guid'] );
```

Comment:
@GrigoreMihai this part seems to repeat pretty often in this PR. We can make it more DRY by moving it into a utility function.